### PR TITLE
Reuse connection pool depending on host and connection options

### PR DESCRIPTION
**Pull request by vhvb1989**
fixes: https://github.com/Azure/azure-sdk-for-cpp/issues/1103
fixes: #1060

### CODE REVIEW

**Review by antkmsft — APPROVED** [review]

**Comment by ahsonkhan** [comment]
> fixes: #1060

No wonder that test would fail occasionally. Keep an eye out for such failures going forward, because they are good canaries for pooling having some bug.

File: sdk/core/azure-core/test/ut/transport_adapter_curl.cpp [hunk 1]
```diff
@@ -36,89 +37,131 @@ namespace Azure { namespace Core { namespace Test {
...
+      Azure::Core::Http::CurlTransportOptions options;
+      auto session = std::make_unique<Azure::Core::Http::CurlSession>(
+          req,
+          Azure::Core::Http::CurlConnectionPool::GetCurlConnection(req, options),
+          options.HttpKeepAlive);
+      // Simulate connection was used already
+      session->m_lastStatusCode = Azure::Core::Http::HttpStatusCode::Ok;
+      session->m_sessionState = Azure::Core::Http::CurlSession::SessionState::STREAMING;
+      EXPECT_EQ(Azure::Core::Http::CurlConnectionPool::ConnectionPoolIndex.size(), 0);
+    }
```
Comment (ahsonkhan) [inline sdk/core/azure-core/test/ut/transport_adapter_curl.cpp:78]: I am trying to understand how this is testing what the comment is suggesting, but haven't difficulty. How are we ensuring the same connection is being used exactly?

File: sdk/core/azure-core/test/ut/transport_adapter_curl.cpp [hunk 2]
```diff
@@ -36,89 +37,131 @@ namespace Azure { namespace Core { namespace Test {
...
-                     "complete."
-                  << std::endl
-                  << "Add compiler option -DRUN_LONG_UNIT_TESTS=OFF when building if you want to "
-                     "skip this test."
-                  << std::endl;
-
-        // Wait for 180 secs to make sure any previous connection is removed by the cleaner
-        std::this_thread::sleep_for(std::chrono::milliseconds(1000 * 180));
-
-        std::cout << "First wait time done. Validating state." << std::endl;
```
Comment (ahsonkhan) [inline sdk/core/azure-core/test/ut/transport_adapter_curl.cpp:89]: Keep this

File: sdk/core/azure-core/test/ut/transport_adapter_curl.cpp [hunk 3]
```diff
@@ -36,89 +37,131 @@ namespace Azure { namespace Core { namespace Test {
...
-
-        // index is not affected by cleaner. It does not remove index
-        EXPECT_EQ(Http::CurlConnectionPool::ConnectionsIndexOnPool(), 1);
-        // cleaner should have removed connections
-        EXPECT_EQ(Http::CurlConnectionPool::ConnectionsOnPool("httpbin.org"), 0);
```
Comment (ahsonkhan) [inline sdk/core/azure-core/test/ut/transport_adapter_curl.cpp:94]: Keep this?

File: sdk/core/azure-core/src/http/curl/curl.cpp [hunk 4]
```diff
@@ -1154,7 +1196,8 @@ void CurlConnectionPool::MoveConnectionBackToPool(
 
   // Lock mutex to access connection pool. mutex is unlock as soon as lock is out of scope
   std::lock_guard<std::mutex> lock(CurlConnectionPool::ConnectionPoolMutex);
-  auto& hostPool = CurlConnectionPool::ConnectionPoolIndex[connection->GetHost()];
+  auto poolId = connection->GetHost() + connection->GetConnectionPropertiesKey();
```
Comment (ahsonkhan) [inline sdk/core/azure-core/src/http/curl/curl.cpp:1199]: Why isn't the `host` included as part of the string returned from `GetConnectionPropertiesKey`? It seems odd to modify the return value again to get the actual pool ID.

**Comment by vhvb1989** [comment]
> No wonder that test would fail occasionally. Keep an eye out for such failures going forward, because they are good canaries for pooling having some bug.

I updated the test in a more unit-test fashion way where there is no network calls. This will ensure the test is stable now and at the same time I am testing the expected bahavior from the pool

File: sdk/core/azure-core/src/http/curl/curl.cpp [hunk 5]
```diff
@@ -1024,7 +1065,8 @@ std::unique_ptr<CurlNetworkConnection> CurlConnectionPool::GetCurlConnection(
     std::lock_guard<std::mutex> lock(CurlConnectionPool::ConnectionPoolMutex);
 
     // get a ref to the pool from the map of pools
-    auto hostPoolIndex = CurlConnectionPool::ConnectionPoolIndex.find(host);
+    auto hostPoolIndex
+        = CurlConnectionPool::ConnectionPoolIndex.find(host + connectionPropertiesKey);
```
Comment (ahsonkhan) [inline sdk/core/azure-core/src/http/curl/curl.cpp:1069]: Can we change `connectionPropertiesKey` to actually be equivalent to `host + connectionPropertiesKey` and not just base it on the options solely?

File: sdk/core/azure-core/src/http/curl/curl.cpp [hunk 6]
```diff
@@ -1012,10 +1012,51 @@ std::map<std::string, std::list<std::unique_ptr<CurlNetworkConnection>>>
 int32_t CurlConnectionPool::s_connectionCounter = 0;
 bool CurlConnectionPool::s_isCleanConnectionsRunning = false;
 
+namespace {
+inline std::string GetConnectionPropertiesKey(CurlTransportOptions const& options)
```
Comment (ahsonkhan) [inline sdk/core/azure-core/src/http/curl/curl.cpp:1016]: These connection string keys **must** be unique and map exactly to 1 option setting with **no collisions**. Come up with an algorithm that guarantees that and is easy to audit and test. There should always be a 1:1 mapping between key string and options (i.e. if I have a key, I can figure out which unique and exactly one option settings it corresponds to, and if I have an option, exactly one unique string can be generated from it).

Currently, two keys that are identical might come from different option values. We need to be careful to avoid that.

Here's an example: "0111"
Now, tell me, what options correspond to that connection key? Is there only a single unique options setting that results in that key?

I am sure there are other possible such strings that cause a collision.

<details><summary>Try it for yourself first, but here's the answer:</summary>
Based on the cainfo and proxy setting strings, I can spoof it and get a connection where keep alive to be on.

0<1>11 <= cainfo is "", keep alive is 0, proxy is "1", norevoke is 1, sslverifypeer is 1
<0>111 <= cainfo is "0", keep alive is 1, proxy is "", norevoke is 1, sslverifypeer is 1
</details>

**Review by ahsonkhan — CHANGES_REQUESTED** [review]
Most importantly: make connection keys unique for any combination of options provided and retain the semantics of the original test, at least for now.

Other than that, consider including the host in the key generation, and some questions/nits. **Keep** the new test as an addition rather than replacing.

File: sdk/core/azure-core/src/http/curl/curl.cpp (hunk 6 above)
Comment (vhvb1989) [reply sdk/core/azure-core/src/http/curl/curl.cpp:1010]: If you are setting CAPath to "0", you are going to have other issues XD.
But I can fix it for sure :)

File: sdk/core/azure-core/src/http/curl/curl.cpp (hunk 5 above)
Comment (vhvb1989) [reply sdk/core/azure-core/src/http/curl/curl.cpp:1063]: Implementation detail.. don't worry about it.

File: sdk/core/azure-core/src/http/curl/curl.cpp (hunk 4 above)
Comment (vhvb1989) [reply sdk/core/azure-core/src/http/curl/curl.cpp:1193]: impl detail

File: sdk/core/azure-core/test/ut/transport_adapter_curl.cpp [hunk 7]
```diff
@@ -36,89 +37,131 @@ namespace Azure { namespace Core { namespace Test {
...
-
-      auto threadRoutine = [&]() {
-        using namespace std::chrono_literals;
-        auto request = Azure::Core::Http::Request(Azure::Core::Http::HttpMethod::Get, host);
-        auto response = m_pipeline->Send(Azure::Core::GetApplicationContext(), request);
-        checkResponseCode(response->GetStatusCode());
-        auto expectedResponseBodySize = std::stoull(response->GetHeaders().at("content-length"));
-        CheckBodyFromBuffer(*response, expectedResponseBodySize);
-        std::this_thread::sleep_for(1s);
-      };
```
[the remark this answers is not in the document]
Comment (vhvb1989) [reply sdk/core/azure-core/test/ut/transport_adapter_curl.cpp:56]: The previous test is not coming back.  It can't be deterministic because it depends on the actual network.
I need to control exactly when a connection is returned to the pool, and by sending a request I can't. It can take nothing to send the request and release the connection, or it could take much more.

File: sdk/core/azure-core/test/ut/transport_adapter_curl.cpp (hunk 2 above)
Comment (vhvb1989) [reply sdk/core/azure-core/test/ut/transport_adapter_curl.cpp:89]: Nop

File: sdk/core/azure-core/src/http/curl/curl.cpp (hunk 5 above)
Comment (ahsonkhan) [reply sdk/core/azure-core/src/http/curl/curl.cpp:1063]: Factors such as code maintenance, ease of understanding, and principles such as single responsibility still matters as part of code review, even if it is just an implementation detail :)

File: sdk/core/azure-core/test/ut/transport_adapter_curl.cpp [hunk 8]
```diff
@@ -36,89 +37,131 @@ namespace Azure { namespace Core { namespace Test {
...
-      EXPECT_EQ(Http::CurlConnectionPool::ConnectionsOnPool("httpbin.org"), 3);
+      // Creating a new connection with default options
+      Azure::Core::Http::CurlTransportOptions options;
+      auto session = std::make_unique<Azure::Core::Http::CurlSession>(
+          req,
+          Azure::Core::Http::CurlConnectionPool::GetCurlConnection(req, options),
+          options.HttpKeepAlive);
+      // Simulate connection was used already
+      session->m_lastStatusCode = Azure::Core::Http::HttpStatusCode::Ok;
+      session->m_sessionState = Azure::Core::Http::CurlSession::SessionState::STREAMING;
```
[the remark this answers is not in the document]
Comment (vhvb1989) [reply sdk/core/azure-core/test/ut/transport_adapter_curl.cpp:61]: When it comes to network, it has been complicated to simulate responses or specific states.
The easiest option is this one where the test can exactly set the next expected state and test some outcome.
The harder option is to add mocked objects where, instead of the test setting this state, we have to model what the mocked object will do.

Note that in both cases, the way we set the specific state is 100% tight coupled to the implementation. That's just how it is XD
Unit tests are meant to test your implementation. If you change your impl, you need to change your tests.
Functional tests are meant to test your public surface. Only when public surface change, the test needs to change

File: sdk/core/azure-core/test/ut/transport_adapter_curl.cpp (hunk 1 above)
Comment (vhvb1989) [reply sdk/core/azure-core/test/ut/transport_adapter_curl.cpp:78]: Before creating the connection we check the number of connections in the pool, which is 1.
Then we create a new session and ask for a new connection to the pool.
Then we check the pool and validate that there are no connections in there. Which means the connection was re-used.

File: sdk/core/azure-core/src/http/curl/curl.cpp (hunk 5 above)
Comment (vhvb1989) [reply sdk/core/azure-core/src/http/curl/curl.cpp]: Well, my general feeling is that it is not such a deal for our team.
I remember I mentioned the `single responsibility` for the winHttp transport adapter when I saw the handle manager as a tight coupling to anyone using it, and you mentioned it's just implementation and that's how you preferred. XD

Anyway, I have done this change :)

Comment (ahsonkhan) [reply sdk/core/azure-core/src/http/curl/curl.cpp]: > when I saw the handle manager as a tight coupling to anyone using it

I had added that based on your recommendation :)

File: sdk/core/azure-core/test/ut/curl_connection_pool.cpp [hunk 9]
```diff
@@ -0,0 +1,172 @@
...
+                ->second.begin()
+                ->get()
+                ->GetConnectionKey(),
+            expectedConnectionKey);
+
+        auto session = std::make_unique<Azure::Core::Http::CurlSession>(
+            req, std::move(connection), options.HttpKeepAlive);
+        // Simulate connection was used already
+        session->m_lastStatusCode = Azure::Core::Http::HttpStatusCode::Ok;
+        session->m_sessionState = Azure::Core::Http::CurlSession::SessionState::STREAMING;
```
Comment (ahsonkhan) [inline sdk/core/azure-core/test/ut/curl_connection_pool.cpp:135]: nit: this test has a bit of code duplication that you might be able to extract into helpers.

File: sdk/core/azure-core/test/ut/curl_connection_pool.cpp [hunk 10]
```diff
@@ -0,0 +1,172 @@
...
+      }
+      // Now there should be 2 index wit one connection each
+      EXPECT_EQ(Azure::Core::Http::CurlConnectionPool::ConnectionPoolIndex.size(), 2);
+      values = Azure::Core::Http::CurlConnectionPool::ConnectionPoolIndex.begin();
+      EXPECT_EQ(values->second.size(), 1);
```
Comment (ahsonkhan) [inline sdk/core/azure-core/test/ut/curl_connection_pool.cpp:140]: What about a test that creates and uses two connections with the same options, and returns it, so that the list is of size 2?

File: sdk/core/azure-core/test/ut/curl_connection_pool.cpp [hunk 11]
```diff
@@ -0,0 +1,172 @@
...
+      EXPECT_EQ(values->second.begin()->get()->GetConnectionKey(), expectedConnectionKey);
+      values++;
+      EXPECT_EQ(values->second.size(), 1);
+      EXPECT_EQ(values->second.begin()->get()->GetConnectionKey(), secondExpectedKey);
+
+      // Test re-using same custom config
+      {
+        // Creating a new connection with default options
+        Azure::Core::Http::CurlTransportOptions options;
+        options.CAInfo = CAinfo;
```
Comment (ahsonkhan) [inline sdk/core/azure-core/test/ut/curl_connection_pool.cpp:118]: We don't have high branch coverage here, because we should also add a test where proxy is non-empty, and one where httpkeepalive is set, so that all the if(true) 1, else 0 branches are covered in the connection key generation.

**Review by ahsonkhan — CHANGES_REQUESTED** [review]
Mainly blocked because I think you missed keepalive.

Other test enhancement (which can be deferred): Add more tests to cover all the key generation branches.

File: sdk/core/azure-core/src/http/curl/curl.cpp [hunk 12]
```diff
@@ -1006,19 +1006,60 @@ std::map<std::string, std::list<std::unique_ptr<CurlNetworkConnection>>>
 int32_t CurlConnectionPool::s_connectionCounter = 0;
 bool CurlConnectionPool::s_isCleanConnectionsRunning = false;
 
+namespace {
+inline std::string GetConnectionKey(std::string const& host, CurlTransportOptions const& options)
+{
+  std::string key(host);
+  if (!options.CAInfo.empty())
```
[the remark this answers is not in the document]
Comment (vhvb1989) [reply sdk/core/azure-core/src/http/curl/curl.cpp:1013]: KeepAlive is not part of the connection settings.
If keepAlive = false, the connection is never re-used anyway.
This setting would never be significant for the pool. 
It is only relevant for the session

[the remark this answers is not in the document]
Comment (ahsonkhan) [reply sdk/core/azure-core/src/http/curl/curl.cpp:1013]: Please add a comment in this method to explain that.

**Review by ahsonkhan — APPROVED** [review]
Other than some comment suggestions and test enhancements, looks good.

File: sdk/core/azure-core/test/ut/curl_connection_pool.cpp (hunk 11 above)
Comment (vhvb1989) [reply sdk/core/azure-core/test/ut/curl_connection_pool.cpp:118]: The number of tests is 2^n, where n is the number of options.
A few combinations should be more than enough.
Otherwise, I would expend more time creating a test generator, but doing it manually would be crazy!:)
